### DISCUSSION TITLE
[GperfTools pperf] How to convert emscripten stack trace to PC and vice-versa

### DISCUSSION THREAD

**Discussion by gmarella** [emscripten-core/emscripten#20008] [Q&A] [answered]
Hi,

I am trying to profile an CPP application (compiled for WASM target) using gperftools heap profiler. I am able to compile the gperftools for WASM target and link it with the application with some changes but currently hit a roadblock with respect to backtrace logic.

### TL;DR
Step 1. How can we get the address (PC) in the backtrace instead of text during profiling?
Step 2. How can we map the address stacktrace to human readable format during analysis(offline) of the collected profiles in step 1?

gperftools internaly expects the backtrace to be presented in an array of addresses(pc's). But emscripten supports backtrace only in the text format. Is there a way to get the actual PCs(or some other format?) as a stacktrace during profiling time and then map these backtraces to human readable text format offline during analyze step.

Thanks in advace.

**Comment by kripken** [comment, marked as answer]
It might help to look at the internals of Emscripten's [sanitizer support](https://emscripten.org/docs/debugging/Sanitizers.html), which does work to convert stack traces to binary offsets. Relevant methods are [emscripten_pc_get_function](https://github.com/emscripten-core/emscripten/blob/9cafed1cb602eda1808232e885b21b4f8d9ae45c/src/library.js#L2725C3-L2725C29), [emscripten_stack_snapshot](https://github.com/emscripten-core/emscripten/blob/9cafed1cb602eda1808232e885b21b4f8d9ae45c/src/library.js#L2628-L2666) and others in that area.

In general, you can convert stack traces to binary offsets. For example, here is part of a stack trace in Node 16:
```js
    at _fd_write (/home/azakai/Dev/emscripten/a.out.js:429:7)
    at __wasi_fd_write (wasm://wasm/85545e82:wasm-function[2]:0x98)
    at __stdio_write (wasm://wasm/85545e82:wasm-function[7]:0x57c)
    at main (wasm://wasm/85545e82:wasm-function[5]:0x4fe)
```
429 is a line number in JS, and 0x98 is a binary offset in wasm. That is parsed by the stack trace handling code I mentioned before.
